phd-tests/tests/src/cpuid.rs

```diff
@@ -0,0 +1,159 @@
+    // defined there.
+    if leaf_0_values.eax < 7 {
+        phd_skip!(
+            "cpuid_boot_test requires the host to report at least leaf 7"
+        );
+    }
+
+    // Keep only standard leaves up to 7 and the first two extended leaves.
+    // Extended leaves 2 through 4 will be overwritten with a fake brand string
+    // (see below).
+    host_cpuid.retain(|leaf, _values| {
+        leaf.leaf <= 0x7
+            || (leaf.leaf >= 0x8000_0000 && leaf.leaf <= 0x8000_0001)
+    });
+
+    // Report that leaf 7 is the last available standard leaf.
+    host_cpuid.get_mut(&CpuidIdent::leaf(0)).unwrap().eax = 7;
+
+    // Mask off feature bits that the Oxide platform won't support. See RFD
+    // 314. These are masks (and not assignments) so that features the host CPU
+    // support won't appear in the guest's feature masks.
+    let leaf_1 = host_cpuid.get_mut(&CpuidIdent::leaf(1)).unwrap();
+    leaf_1.ecx &= 0xF6F8_3203;
+    leaf_1.edx &= 0x178B_FBFF;
+    let leaf_7 = host_cpuid.get_mut(&CpuidIdent::leaf(7)).unwrap();
+    leaf_7.eax = 0;
+    leaf_7.ebx &= 0x219C_03A9;
+    leaf_7.ecx = 0;
+    leaf_7.edx = 0;
+
+    // Report that leaf 0x8000_0004 is the last extended leaf and clean up
+    // feature bits in leaf 0x8000_0001.
+    host_cpuid.get_mut(&CpuidIdent::leaf(0x8000_0000)).unwrap().eax =
+        0x8000_0004;
+    let ext_leaf_1 =
+        host_cpuid.get_mut(&CpuidIdent::leaf(0x8000_0001)).unwrap();
+    ext_leaf_1.ecx &= 0x4440_01F0;
+    ext_leaf_1.edx &= 0x27D3_FBFF;
+
+    // Test the plumbing by pumping a fake processor brand string into extended
+    // leaves 2-4 and seeing if the guest recognizes it.
+    const BRAND_STRING: &[u8; 48] =
+        b"Oxide Cloud Computer Company Cloud Computer\0\0\0\0\0";
+
+    let chunks = BRAND_STRING.chunks_exact(4);
+    let mut ext_leaf_2 = CpuidValues::default();
+    let mut ext_leaf_3 = CpuidValues::default();
+    let mut ext_leaf_4 = CpuidValues::default();
+    let dst = ext_leaf_2
+        .iter_mut()
+        .chain(ext_leaf_3.iter_mut())
+        .chain(ext_leaf_4.iter_mut());
+
+    for (chunk, dst) in chunks.zip(dst) {
+        *dst = u32::from_le_bytes(chunk.try_into().unwrap());
+    }
+
+    host_cpuid.insert(CpuidIdent::leaf(0x8000_0002), ext_leaf_2);
+    host_cpuid.insert(CpuidIdent::leaf(0x8000_0003), ext_leaf_3);
+    host_cpuid.insert(CpuidIdent::leaf(0x8000_0004), ext_leaf_4);
+
+    // Try to boot a guest with the computed CPUID values. The modified brand
+    // string should show up in /proc/cpuinfo.
+    let mut cfg = ctx.vm_config_builder("cpuid_boot_test");
+    cfg.cpuid(
+        host_cpuid
+            .iter()
+            .map(|(leaf, value)| CpuidEntry {
+                leaf: leaf.leaf,
+                subleaf: leaf.subleaf,
+                eax: value.eax,
+                ebx: value.ebx,
+                ecx: value.ecx,
+                edx: value.edx,
+            })
+            .collect(),
```

Comment:
Hmm, I thought I had a `From` impl for this sort of thing...

Reply:
Done (hopefully without creating too many headaches for myself when I rebase #782).

Reply:
Oh, right, I forgot: PHD is pulling in the generated type from propolis_client here, which doesn't get the From impls from propolis_api_types. This is why the build broke; I'll back this out to patch it up.